### PR TITLE
Hotfix url array params

### PR DESCRIPTION
**Pull request by josemato**
Add support for http arrays on query string and maintain consistency between variable names. Example:

1) with chaplin this url: domain.com/?foo=hi&foo=guys&bar=1 will parse to next object: {foo: [hi, guys], bar: 1}

2) same urls using arrays syntax with brackets:  domain.com/?foo[]=hi&foo[]=guys&bar=1 will parse to next object: {"foo[]": [hi, guys], bar: 1}

So, the problem is that chaplin at this moment is not treating array synxtax with brackets at the same way like array syntax in step 1. With this fix, parse object will be the same as step 1: {foo: [hi, guys], bar: 1}

=> Other scenario will be when you want that "foo" variable will be an array on your application but when this variable has only one element then react will think that is not an array because you call the url with only one value but in other request could be two and then the app will not consistence.

### CODE REVIEW

File: src/chaplin/lib/utils.coffee [hunk 1]
```diff
@@ -134,6 +134,8 @@ utils =
         continue unless field.length
         field = decodeURIComponent field
         value = decodeURIComponent value
+
+        field = field.substring(0, field.length - 2) if field.indexOf('[]') == field.length - 2 && field.length - 2 > -1
```
Comment (paulmillr) [inline src/chaplin/lib/utils.coffee]:
```coffeescript
len = field.length
field = field.substring(0, len - 2) if len > 2 and field.indexOf('[]') is len - 2
```

**Comment by paulmillr** [comment]
Thanks for the patch. Please don't forget add the tests.

**Comment by josemato** [comment]
Thi @paulmillr I had problems with test on my computer running "npm test" getting next error:
Warning: Task "mocha" not found.

I don't know how could i do the tests, could you explain me? Thanks

**Comment by akre54** [comment]
Theres a million and one different ways to parse a url (and even more when dealing with arrays). Some people use commas to delimit values, for example. Can we make this a user-exposed and overridable feature instead of baking this particular implementation into Chaplin itself?

**Comment by josemato** [comment]
Hi @akre54 Nowadays libraries like jquery support this two methods for work ("traditional" => array=1&array=2 and "modern way" => array[]=1&array[]=2").
I created a couple of html forms and browsers only can send this data using "traditional" or "modern" way, no more (no comma separated by default for example):
http://pastebin.com/i0Dru4wY

In the other hand, php, ruby on rails support "modern way", "java" supports traditional way and nodejs support boh.

I think this is a needed change on chaplin because a frontend framework will need to work hand to hand with backend and supports similar cases (else it's a mess do enconding from backend to frontend and viceversa). It will be great receive params on controller in a transparent way.
Thanks.

**Comment by vendethiel** [comment]
Maybe Chaplin should offer implementations, and the user can decide which one to choose (or use a custom function)

**Comment by josemato** [comment]
Yes @vendethiel It would be a solution for this problem. But, there are two implementations (support by browsers) so maybe it could be a transparent solution for parser function, anyway, is your choice :)

**Comment by paulmillr** [comment]
Seems like we have too many users who'll disagree with this particular implementation. Sorry for that.

**Comment by josemato** [comment]
Your are right @paulmillr but while we "discuss" this issue chaplin still not supports arrays in urls in anyway (the example is when you need an array variable with only one value depends on the situation => chaplin thinks it's a primitive variable) :(

**Comment by askmatey** [comment]
Using brackets in a url is **bad coding**. Your browser can handle the query parameter by accident.

Your query param `?foo[]=hi&foo[]=guys&bar=1` is **not valid**, see [RFC 3986](http://tools.ietf.org/html/rfc3986):

> URIs include components and subcomponents that are delimited by characters in the "reserved" set. These characters are called "reserved" because they may (or may not) be defined as delimiters by the generic syntax, by each scheme-specific syntax, or by the implementation-specific syntax of a URI's dereferencing algorithm.

>        reserved    = gen-delims / sub-delims
       gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"
       sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                   / "*" / "+" / "," / ";" / "="

https://stackoverflow.com/questions/2366260/whats-valid-and-whats-not-in-a-uri-query#answer-2375597

**Comment by josemato** [comment]
I am agree with you @askmatey  but you can encode url params (for example using php function http_build_query https://php.net/http_build_query) and you can see that chaplin not understand array parameters.

**Comment by askmatey** [comment]
Each web environment provides its own 'standard' for such things.

Just because php can interpret the brackets as array, does not mean they are not first-class citizens anyway.

**Comment by josemato** [comment]
Yes @askmatey  but that's not the point. The point is, ¿which is the way to pass an array in url and that data could be parsed by chaplin (array could be only one element sometimes)? No exists any way in chaplin to do that.

**Pull request closed without merging**